core/state/database.go

```diff
@@ -238,19 +238,57 @@ func (db *cachingDB) purgeLoop() {
 
 // OpenTrie opens the main account trie at a specific root hash.
 func (db *cachingDB) OpenTrie(root common.Hash) (Trie, error) {
+	//try cache first
+	if db.accountTrieCache != nil {
+		if tr, ok := db.accountTrieCache.Get(root); ok {
+			return tr.(*trie.SecureTrie), nil
+		}
+	}
+
 	tr, err := trie.NewStateTrie(trie.StateTrieID(root), db.triedb)
 	if err != nil {
 		return nil, err
 	}
+
+	//update LRU
+	if db.accountTrieCache != nil {
+		db.accountTrieCache.Add(root, tr.ResetCopy())
+	}
 	return tr, nil
 }
 
 // OpenStorageTrie opens the storage trie of an account.
 func (db *cachingDB) OpenStorageTrie(stateRoot common.Hash, addrHash, root common.Hash) (Trie, error) {
+	// try cache first
+	if db.storageTrieCache != nil {
+		if tries, exist := db.storageTrieCache.Get(addrHash); exist {
+			for _, triePair := range tries {
+				if triePair.root == root {
+					return triePair.trie, nil
+				}
+			}
+		}
+	}
+
 	tr, err := trie.NewStateTrie(trie.StorageTrieID(stateRoot, addrHash, root), db.triedb)
 	if err != nil {
 		return nil, err
 	}
+
+	if db.storageTrieCache != nil {
+		tries, exist := db.storageTrieCache.Get(addrHash)
+		var newTriesArray [3]*triePair
+		if exist {
+			newTriesArray = [3]*triePair{
+				{root: root, trie: tr.ResetCopy()},
+				tries[0],
+				tries[1],
+			}
+		} else {
+			newTriesArray = [3]*triePair{{root: root, trie: tr.ResetCopy()}, nil, nil}
+		}
+		db.storageTrieCache.Add(addrHash, newTriesArray)
+	}
```

Comment:
```suggestion

	db.CacheStorage(addrHash, root, tr)
```

How about reusing `db.CacheStorage()` to update the cache?